From 0cbc42f4dc6b3763532176b67141aac0c8a400d7 Mon Sep 17 00:00:00 2001
From: Haoran Zhang <haoranzhang@google.com>
Date: Thu, 12 Jun 2025 09:33:17 -0700
Subject: [PATCH] Implement onNullBinding() in autofill service connection

With this fix, an autofill service which returns null in onBind() will no longer stay bound by system_server and will no longer be able to launch activities from the background.

Flag: EXEMPT security bugfix
Test: sts-tradefed run sts-dynamic-develop -m StsHostTestCases -t android.security.sts.Bug_401545800
Bug:b/401545800
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:d883efa912c0a39c46437074576cffb7078ac455)
Merged-In: Ieb3e63fe2ac1f440be3d164730ef3110209ec1a6
Change-Id: Ieb3e63fe2ac1f440be3d164730ef3110209ec1a6
---
 .../com/android/server/autofill/RemoteFillService.java    | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/services/autofill/java/com/android/server/autofill/RemoteFillService.java b/services/autofill/java/com/android/server/autofill/RemoteFillService.java
index 07f5dcc3cb0a..76cc8234d0b8 100644
--- a/services/autofill/java/com/android/server/autofill/RemoteFillService.java
+++ b/services/autofill/java/com/android/server/autofill/RemoteFillService.java
@@ -116,6 +116,14 @@ final class RemoteFillService extends ServiceConnector.Impl<IAutoFillService> {
         }
     }
 
+    @Override // from ServiceConnection
+    public void onNullBinding(@NonNull ComponentName name) {
+        if (sVerbose) {
+            Slog.v(TAG, "onNullBinding");
+        }
+        unbind();
+    }
+
     private void dispatchCancellationSignal(@Nullable ICancellationSignal signal) {
         if (signal == null) {
             return;
-- 
2.50.1.351.gcf940e82a1-goog

